internal/resources/providers/gcplib/inventory/provider.go

```diff
@@ -91,242 +75,327 @@ type dnsPolicyFields struct {
+	policyCh := make(chan *assetpb.Asset)   // *assetpb.Asset with IamPolicy
+	mergeCh := make(chan *assetpb.Asset)    // *assetpb.Asset with Resource and IamPolicy
+	enrichCh := make(chan *ExtendedGcpAsset)
 
-	var assets []*assetpb.Asset
-	assets = append(append(assets, resourceAssets...), policyAssets...)
-	mergedAssets := mergeAssetContentType(assets)
-	extendedAssets := p.extendWithCloudMetadata(ctx, mergedAssets)
-	// Enrich network assets with dns policy
-	p.enrichNetworkAssets(ctx, extendedAssets)
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_RESOURCE, assetTypes, resourceCh)
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_IAM_POLICY, assetTypes, policyCh)
+	go p.mergeAssets(ctx, mergeCh, resourceCh, policyCh)
+	go p.enrichAssets(ctx, mergeCh, enrichCh)
 
-	return extendedAssets, nil
+	for asset := range enrichCh {
+		out <- asset
+	}
 }
 
-// ListMonitoringAssets returns a list of monitoring assets grouped by project id
-func (p *Provider) ListMonitoringAssets(ctx context.Context, monitoringAssetTypes map[string][]string) ([]*MonitoringAsset, error) {
-	logMetrics, err := p.ListAllAssetTypesByName(ctx, monitoringAssetTypes["LogMetric"])
-	if err != nil {
-		return nil, err
+func (p *Provider) ListMonitoringAssets(ctx context.Context, out chan<- *MonitoringAsset) {
+	defer close(out)
+
+	projectsCh := make(chan *assetpb.Asset)
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_RESOURCE, []string{CrmProjectAssetType}, projectsCh)
+
+	for project := range projectsCh {
+		logsResourceCh := make(chan *assetpb.Asset)
+		alertsResourceCh := make(chan *assetpb.Asset)
+		logsAssetCh := make(chan *ExtendedGcpAsset)
+		alertsAssetCh := make(chan *ExtendedGcpAsset)
+
+		go p.getAllAssets(ctx, project.Ancestors[0], assetpb.ContentType_RESOURCE, []string{MonitoringLogMetricAssetType}, logsResourceCh)
+		go p.getAllAssets(ctx, project.Ancestors[0], assetpb.ContentType_RESOURCE, []string{MonitoringAlertPolicyAssetType}, alertsResourceCh)
+		go p.enrichAssets(ctx, logsResourceCh, logsAssetCh)
+		go p.enrichAssets(ctx, alertsResourceCh, alertsAssetCh)
+
+		var logAssets, alertAssets []*ExtendedGcpAsset
+		var wg sync.WaitGroup
+
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			logAssets = collect(logsAssetCh)
+			p.log.Debugf("Listed %d log metrics for %v\n", len(logAssets), project.Name)
+		}()
+		go func() {
+			defer wg.Done()
+			alertAssets = collect(alertsAssetCh)
+			p.log.Debugf("Listed %d alert policies for %v\n", len(alertAssets), project.Name)
+		}()
+		wg.Wait()
+		out <- &MonitoringAsset{
+			LogMetrics:   logAssets,
+			Alerts:       alertAssets,
+			CloudAccount: logAssets[0].CloudAccount,
+		}
 	}
+}
 
-	alertPolicies, err := p.ListAllAssetTypesByName(ctx, monitoringAssetTypes["AlertPolicy"])
-	if err != nil {
-		return nil, err
-	}
+func (p *Provider) ListProjectsAncestorsPolicies(ctx context.Context, out chan<- *ProjectPoliciesAsset) {
```

Comment:
this used to send a slice of items that each represent a group of assets of the same project. it now does the same, but sends each group separately so we don't wait for all of them.